salt/modules/vagrant.py

```diff
@@ -0,0 +1,466 @@
+    reply = __salt__['cmd.shell'](cmd, cwd)
+    for line in reply.split('\n'):  # build a list of the text reply
+        print(line)
+        tokens = line.strip().split()
+        if len(tokens) > 1 and tokens[-1].endswith(')'):
+            try:
+                info[tokens[0]] = {'state': ' '.join(tokens[1:-1])}
+                info[tokens[0]]['provider'] = tokens[-1].lstrip('(').rstrip(')')
+            except IndexError:
+                pass
+    return info
+
+
+def init(name,  # Salt_id for created VM
+         cwd=None,   # path to find Vagrantfile
+         machine='',  # name of machine in Vagrantfile
+         runas=None,  # username who owns Vagrant box
+         start=False,  # start the machine when initialized
+         deploy=None,  # flag suggesting whether to load Salt onto the VM
+         vagrant_provider='',  # vagrant provider (default=virtualbox)
+         vm=None,  # a dictionary of VM configuration settings
+         ):
+    '''
+    Initialize a new Vagrant vm
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt <host> vagrant.init <salt_id> /path/to/Vagrantfile
+        salt my_laptop vagrant.init x1 /projects/bevy_master machine=quail1
+    '''
+    vm_ = {} if vm is None else vm.copy()  # passed configuration data
+    vm_['name'] = name
+    # passed-in keyword arguments overwrite vm dictionary values
+    vm_['cwd'] = cwd or vm_.get('cwd')
+    if not vm_['cwd']:
+        raise SaltInvocationError('Path to Vagrantfile must be defined by \'cwd\' argument')
+    vm_['machine'] = machine or vm_.get('machine', machine)
+    vm_['runas'] = runas or vm_.get('runas', runas)
+    vm_['deploy'] = deploy if deploy is not None else vm_.get('deploy', True)
+    vm_['vagrant_provider'] = vagrant_provider or vm_.get('vagrant_provider', '')
+    _update_cache(name, vm_)
+
+    if start:
+        log.debug('Starting VM {0}'.format(name))
+        ret = _start(name, vm_)
+    else:
+        ret = 'Name {} defined using VM {}'.format(name, vm_['machine'] or '(default)')
+    return ret
+
+
+def start(name):
+    '''
+    Start (vagrant up) a defined virtual machine by salt_id name.
+    The machine must have been previously defined using "vagrant.init".
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt <host> vagrant.start <salt_id>
+    '''
+    vm_ = get_vm_info(name)
+    return _start(name, vm_)
+
+
+def _start(name, vm_):  # internal call name, because "start" is a keyword argument to vagrant.init
+
+    try:
+        machine = vm_['machine']
+    except KeyError:
+        raise SaltInvocationError('No Vagrant machine defined for Salt-id {}'.format(name))
+
+    vagrant_provider = vm_.get('vagrant_provider', '')
+    provider_ = '--provider={}'.format(vagrant_provider) if vagrant_provider else ''
+    cmd = 'vagrant up {} {}'.format(machine, provider_)
+    ret = __salt__['cmd.retcode'](cmd, runes=vm_.get('runas'), cwd=vm_.get('cwd'))
+
+    return ret == 0
+
+
+def shutdown(name):
```

Comment:
We could just do a `__func_alias__` here, I think. See other modules for an example.

Reply:
That's a nifty trick! I was unaware of it.

I tried the suggestion.  I converted the signature to `start_(machine, vm_=None)` and then filled in the vm_ when it was null. 

After `make html`, the resulting html document contains the reference to the optional `vm_` argument.

That argument is useful for internal calls only.  I could not figure out a good way to explain that in the docs, so I reverted the code to maintain the separation between the external and internal calls.  I think the cleaner documentation is worth the cost of the extra call in this case.